Use Iceberg Api instead of Spark

jcellary opened the pull request:
Replace spark with direct pyiceberg usage.

jcellary committed:
wip

jcellary committed:
fix

jcellary committed:
fix

joaopamaral commented on target_iceberg/sinks.py:
```diff
@@ -26,13 +23,12 @@ def __init__(self, target, schema, stream_name, key_properties) -> None:
         table_name_prefix = f"{self.config.get('table_name_prefix')}_" if self.config.get("table_name_prefix") else ""
-        self.table_name = f"{self.config['db_name'] if self.config.get('prod') else 'scratch'}.{table_name_prefix}{self.__class__.to_snake_case(self.stream_name)}"
+        self.table_name = (f"{self.config['db_name'] if self.config.get('prod') else 'scratch'}."
+                           f"{table_name_prefix}{self.__class__.to_snake_case(self.stream_name)}")
```
```suggestion
                           f"{table_name_prefix}{IcebergSink.to_snake_case(self.stream_name)}")
```

joaopamaral commented on target_iceberg/utils.py:
```diff
@@ -0,0 +1,102 @@
+logger = logging.getLogger(__name__)
+
+
+def convert_size_to_bytes(size_str: str) -> int:
+    """Convert a size string to bytes."""
+    match = re.match(r"(\d+)([KkMmGg]?)", size_str)
+    if not match:
+        raise ValueError(f"Invalid size string: {size_str}")
+
+    value, unit = match.groups()
+    value = int(value)
+
+    if unit.lower() == "k":
+        value *= 1024
+    elif unit.lower() == "m":
+        value *= 1024 * 1024
+    elif unit.lower() == "g":
+        value *= 1024 * 1024 * 1024
+    else:
+        raise ValueError(f"Invalid unit: {unit}")
+
+    return value
```
```suggestion
```

joaopamaral commented on target_iceberg/utils.py:
```diff
@@ -0,0 +1,102 @@
+    input_type = next(iter(types_uppercase)) if types_uppercase else ""
+    pyarrow_type = FIELD_TYPE_TO_PYARROW.get(input_type, pa.string())
+    # override with timestamp type if format set to date or date-time even if value type is e.g. string.
+    if input_types.get("format") in ["date", "date-time"]:
+        pyarrow_type = pa.timestamp("ms")
```
Let's add a test to validate this part.

joaopamaral reviewed (approved):
Just left a few comment but LGTM.

Khrol commented on target_iceberg/utils.py:
```diff
@@ -0,0 +1,102 @@
+def create_pyarrow_table(list_dict: list[dict], schema: pa.Schema) -> pa.Table:
+    """Create a pyarrow Table from a python list of dict."""
+    data = {f: [row.get(f) for row in list_dict] for f in schema.names}
+    return pa.table(data).cast(schema)
```
Configure IDE to auto-apply the last new line.

Khrol commented on target_iceberg/utils.py:
```diff
@@ -0,0 +1,102 @@
+    return value
+
+
+def _field_type_to_pyarrow_field(
```
Unit-test it, please. Lots of things happen here.

jcellary committed:
test

jcellary committed:
fix

jcellary committed:
Update target_iceberg/sinks.py

Co-authored-by: Joao Amaral <7281460+joaopamaral@users.noreply.github.com>

jcellary committed:
Update target_iceberg/utils.py

Co-authored-by: Joao Amaral <7281460+joaopamaral@users.noreply.github.com>

jcellary committed:
added test

jcellary commented:
Addressed everything

Khrol commented on target_iceberg/target.py:
```diff
@@ -3,6 +3,7 @@
 from __future__ import annotations
 
 import json
+import sys
```
?

jcellary replied:
it's used [here](https://github.com/Automattic/target-iceberg/pull/8/files#diff-0e2eabaefccddcc7156f7802024c56809e4a5096a32273a985494f7591d767e4L71).

Khrol replied:
I see. Basically bug fix.

joaopamaral commented on target_iceberg/target.py:
```diff
@@ -3,6 +3,7 @@
 import json
+import sys
 from singer_sdk import typing as th
 from singer_sdk.target_base import Target
```
@jcellary, I think we can add a config to inform the target the schemas that will overwrite the data (as the replication method is not currently available in target https://github.com/meltano/sdk/discussions/3071)

jcellary committed:
overwrite/append

jcellary committed:
style fix

jcellary commented:
I addressed all comments and added support for append and overwrite modes. They are controlled by a config flag.

I also looked into replacing synced_ms with https://sdk.meltano.com/en/latest/implementation/record_metadata.html, however after putting some more thought into it, I'm not convinced that's a good idea, since that adds 6 extra columns and I don't see any use cases for the other 5. So while it would be nice to have sync time added automatically instead of explicitly in code I think these few lines are the lesser evil from 5 extra unneeded columns.

jcellary commented:
overwrite flag is set in airflow-ext [here](https://github.com/Automattic/nosara/pull/11261/files#diff-b086deaa8fd946e5f8b95aca726bda606f98dfc6660c161d3276971dbbc91209R160).

joaopamaral commented on target_iceberg/sinks.py:
```diff
@@ -73,88 +72,40 @@ def max_size(self) -> int:
+    def clean_up(self) -> None:
+        """Perform any clean up actions required at end of a stream."""
+        if self.overwrite_data:
+            self.table.overwrite(self.data_buffer)
```
nit
```suggestion
            self.logger.info(f'Overwriting data in the table {self.table_name}')
            self.table.overwrite(self.data_buffer)
```

joaopamaral commented on target_iceberg/sinks.py:
```diff
@@ -73,88 +72,40 @@ def max_size(self) -> int:
+        """Perform any clean up actions required at end of a stream."""
+        if self.overwrite_data:
+            self.table.overwrite(self.data_buffer)
+        super().clean_up()
```
```suggestion
        super().clean_up()

```

joaopamaral reviewed (approved):
LGTM! :shipit:

joaopamaral commented on target_iceberg/sinks.py:
```diff
@@ -26,13 +24,14 @@ def __init__(self, target, schema, stream_name, key_properties) -> None:
+                           f"{table_name_prefix}{IcebergSink.to_snake_case(self.stream_name)}")
         self.flatten_max_level = self.config.get("max_flatten_level", 0)
         self.skip_add_synced_field = self.config.get("skip_add_synced_field", False)
+        self.overwrite_data = self.config.get("overwrite_data", False)
```
It can be boolean as we can have incremental streams and full sync running with the same command. Let's make it a list of streams.
```suggestion
        self.overwrite_data = self.config.get("overwrite_data", '').split(',')
```

joaopamaral reviewed (changes requested):
Sorry, just noticed that overwrite_data can't be a boolean as we the streams share the same config and we can have incremental and full table with the same config.

jcellary committed:
Update target_iceberg/sinks.py

Co-authored-by: Joao Amaral <7281460+joaopamaral@users.noreply.github.com>

jcellary committed:
Update target_iceberg/sinks.py

Co-authored-by: Joao Amaral <7281460+joaopamaral@users.noreply.github.com>

jcellary committed:
Update target_iceberg/sinks.py

Co-authored-by: Joao Amaral <7281460+joaopamaral@users.noreply.github.com>

jcellary committed:
introduced overwrite_data_for_streams

joaopamaral reviewed (approved):
LGTM! 🚀
